### DISCUSSION TITLE
grid_enum potential TODOs

### DISCUSSION THREAD

**Discussion by alwinm** [cholla-hydro/cholla#215] [Ideas]
I'm listing a number of possible TODOs related to grid_enum. In the short-term, I was focused on increasing the extensibility of SCALARS, but I also wanted to make sure that other things could be refactored along these lines piecemeal on an opt-in basis.

I sort this TODO list roughly in order of priority as I view it, placing low effort high value things closer the top. 

- [ ] - New fields not currently in dev or main (potentially high priority if grid_enum breaks them)
- [ ] - use grid_enum::num_fields to set H.nfields
- [ ] - GasEnergy (instead of num_fields - 1)
- [x] - Magnetic fields (magnetic_x/y/z, instead of NSCALARS + 5,6,7) (#220)
- [ ] - Easy Hydro variables (density, momentum_x/y/z, Energy) (high effort)
- [ ] - Difficult Hydro variables 

Hydro variables has two parts because some are easy cases (0*ncells, 1*ncells, etc.) but some are a little fancier:
here is an example:

https://github.com/cholla-hydro/cholla/blob/d2a851dd5d64b3aef43197d084c0d6858ff0c639/src/reconstruction/plmc_cuda.cu#L117

In the above example, I think it'd be fine to replace the integers in:
https://github.com/cholla-hydro/cholla/blob/d2a851dd5d64b3aef43197d084c0d6858ff0c639/src/reconstruction/plmc_cuda.cu#L27

but every such "difficult" example is more difficult than a drop-in replace. 

The two main constraints which I think are more trouble than they're worth to get around, are:
- keeping scalars in a consecutive bundle (because evolving passive scalars rely on for loops through the scalars)
- keeping the default hydro variables where they are
Aside from these main constraints, in the final form of grid_enum, it should be possible to re-order variables and add new ones to the list without as much thought/work as currently needed.

**Comment by alwinm** [comment]
Move more calls to C.host and C.device instead of the conserved array names.

**Comment by alwinm** [comment]
On the subject of I/O, we should discuss potentially moving away from the 'for loop' over scalars ```scalar0 scalar1 ... scalarN``` and requiring each scalar be named according to its physics name. Since currently whether or not the scalar loop is called is surrounded by ifdefs for CHEMISTRY.

Both for input (restarting) and output.

**Comment by alwinm** [comment]
Locating any residual NSCALARS +:

```
git grep -ip NSCALARS[\s]*+
git grep -ip +[\s]*NSCALARS
```

**Comment by bcaddy** [comment]
As an FYI. Some of the computations for the constrained transport electric fields do depend on field ordering and can't be easily changed to use grid enums. They could be transformed but it would probably require converting `mhd::_internal::_ctSlope` into a template to maintain performance. The template parameters would be the various shifts so that they're determined at compile time rather than causing lots of warp divergence.

**Reply by bcaddy** [reply]
Scratch that. I think I fixed it as long as the momenta fields stay right next to each other.
